### ISSUE TITLE
Build fails with experimental ESM loader

### ISSUE THREAD

**Issue by kherock** [vercel/next.js#31279]
### What version of Next.js are you using?

12.0.3

### What version of Node.js are you using?

16.13.0

### What browser are you using?

N/A

### What operating system are you using?

macOS

### How are you deploying your application?

N/A

### Describe the Bug

When using Next.js with Yarn 3's experimental ESM loader enabled, builds will hang for a while and eventually fail. The build process's dependence on jest-worker is causing builds to fail when `--experimental-loader` is passed as a Node runtime argument. There isn't much that Next.js can do to fix this, but I just wanted to bring attention to this issue since PnP users will otherwise have to explicitly opt-out of ESM support starting in Yarn v3.2 and Next.js 12.

I've opened an issue on jest-worker tracked here: https://github.com/facebook/jest/issues/12060

I want to make it clear that this is **not** a Yarn/PnP bug. This is probably a Node.js quirk with how they've implemented their ESM loader hook.

### Expected Behavior

The app should build a starter project without errors with Yarn 3 while PnP is enabled.

### To Reproduce

A reproduction repo is available here: https://github.com/kherock/jest-worker-spawn-bug

Please check out the README there for more context.

```console
$ yarn install
$ yarn build-with-loader
node:internal/validators:119
    throw new ERR_INVALID_ARG_TYPE(name, 'string', value);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "id" argument must be of type string. Received null
    at new NodeError (node:internal/errors:371:5)
    at validateString (node:internal/validators:119:11)
    at Module.require (node:internal/modules/cjs/loader:998:3)
    at require (node:internal/modules/cjs/helpers:102:18)
    at execMethod (/Users/herockk/Workspaces/jest-worker-spawn-bug/node_modules/jest-worker/build/workers/processChild.js:107:16)
    at process.messageListener (/Users/herockk/Workspaces/jest-worker-spawn-bug/node_modules/jest-worker/build/workers/processChild.js:39:7)
    at process.emit (node:events:390:28)
    at emit (node:internal/child_process:917:12)
    at processTicksAndRejections (node:internal/process/task_queues:84:21) {
  code: 'ERR_INVALID_ARG_TYPE'
}
```

**Comment by kherock** [comment]
I've just realized that we can use `worker_threads` as a workaround, it doesn't have this timing issue. In jest-worker this can be enabled by passing `enableWorkerThreads: true` to the worker options. Worker threads are available in all maintained Node versions:

https://nodejs.org/docs/latest-v12.x/api/worker_threads.html

Jest also implements a fallback in case the current environment doesn't support them.

**Comment by sidhuko** [comment]
Maybe a FAQ, I believe it should work as normal if you set `pnpEnableEsmLoader: false` in your .yarnrc.yml file.

**Comment by kherock** [comment]
> Maybe a FAQ, I believe it should work as normal if you set `pnpEnableEsmLoader: false` in your .yarnrc.yml file.

I depend on ESM packages, this isn't an option for me. However this issue is fixed for me with Node 16.14, so I'm going to close this.
